### PR TITLE
siddon revised

### LINKED ISSUE #266: implement `grid_sample` in `renderers.py` and in-place operations
Siddon's method can be improved by implementing `grid_sample`. Since `grid_sample`  could be used in masking and trilinear method, a function to calculate voxels and rays is beneficial. 

Additionally, memory management could be improved by implementing in-place tensor operations such as `mul_` and `div_`.

### PR DESCRIPTION
**Pull request by momen-mh**
closes #266 

Has problems with optimizers, which need to be addressed

### CODE REVIEW

File: diffdrr/renderers.py
Comment (eigenvivek) [inline diffdrr/renderers.py:1]: the `optimizers` notebook failing is because of the `nansum` calls

Comment (momen-mh) [reply diffdrr/renderers.py:1]: I kept the `nansum` since `step_length = torch.diff(alphas, dim=-1)` generates `nan`s. Would a better solution be to use `nan_to_num()` and then use `sum`?

Comment (eigenvivek) [reply diffdrr/renderers.py:1]: I think we can actually get rid of the `get_alpha_minimax` function entirely. I'll play around with it and push an update to this branch once I figure it out.

Comment (eigenvivek) [reply diffdrr/renderers.py:1]: After removing the `nan` ops, Siddon's method implemented with `grid_sample` returns gradients. However, these gradients don't seem to be useful for camera pose optimization. This is strange since the existing implementation, which is effectively a pure Python implementation of `grid_sample`, works perfectly.

Perhaps the `grid_sample` call needs a custom backwards function?

Comment (eigenvivek) [reply diffdrr/renderers.py:1]: Found a few MINISCULE mistakes in the parameterization of the planes / computation of intersections / normalizing of rays (boiled down to needing to subtract 0.5 some places / subtracting 1 in other places). Anyways, implemented the changes and now `optimizers.ipynb` runs. Also reran all the other tutorials. PR is ready for a final review.

**Comment by eigenvivek** [comment]
Double check if `_get_voxels` needs to be wrapped with `torch.no_grad()`... in the original implementation, casting indices to `long()` removes the gradient. Confirm that `grid_sample` with `mode="nearest"` achieves the same effect.

https://github.com/momen-mh/DiffDRR/blob/3dd9526392ac8965544c774b3bc25cb50fef0043/diffdrr/renderers.py#L51-L52

**Comment by momen-mh** [comment]
Also double check if in-place operations are the issue.
